### PR TITLE
Extract cluster management for integration tests into JUnit test rule out of OpenSearchIntegTestCase

### PR DESCRIPTION
**Pull request by reta**
### Description
Extract cluster management for integration tests into JUnit test rules our of OpenSearchIntegTestCase. @neetikasinghal this is follow up on the work you have done for parameterizing integration test cases. We now have removed the test cluster management logic from the `OpenSearchIntegTestCase` to dedicated JUnit rule, which gave us an opportunity implement two favours of cluster setting parametrization: 

- `ParameterizedDynamicSettingsOpenSearchIntegTestCase` (was implemented by `ParameterizedOpenSearchIntegTestCase` ): the test clusters settings are applied dynamically before each test method
- `ParameterizedStaticSettingsOpenSearchIntegTestCase` (was not supported): the clusters settings are applied at startup time 

CC @andrross should open up the road for parameterized segment replication test cases

### Related Issues
N/A

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
> ❌ Gradle check result for [1739955](https://github.com/opensearch-project/OpenSearch/commit/1739955b266613330cf2103b8b4412c3cf78bd51): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32176/)

https://github.com/opensearch-project/OpenSearch/issues/10006

**Comment by reta** [comment]
> ❌ Gradle check result for [1739955](https://github.com/opensearch-project/OpenSearch/commit/1739955b266613330cf2103b8b4412c3cf78bd51): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32177/)

https://github.com/opensearch-project/OpenSearch/issues/9191
https://github.com/opensearch-project/OpenSearch/issues/10749
https://github.com/opensearch-project/OpenSearch/issues/10755

**Comment by reta** [comment]
> ❌ Gradle check result for [e428d4c](https://github.com/opensearch-project/OpenSearch/commit/e428d4c394d6f31167eee677ca5d0fb3ad57eaf1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32201/)

https://github.com/opensearch-project/OpenSearch/issues/9191
https://github.com/opensearch-project/OpenSearch/issues/10983
https://github.com/opensearch-project/OpenSearch/issues/10154

**Comment by reta** [comment]
> ❌ Gradle check result for [e428d4c](https://github.com/opensearch-project/OpenSearch/commit/e428d4c394d6f31167eee677ca5d0fb3ad57eaf1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32202/)

https://github.com/opensearch-project/OpenSearch/issues/9191

**Comment by reta** [comment]
> ❌ Gradle check result for [e428d4c](https://github.com/opensearch-project/OpenSearch/commit/e428d4c394d6f31167eee677ca5d0fb3ad57eaf1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32207/)

https://github.com/opensearch-project/OpenSearch/issues/1703

File: modules/analysis-common/src/internalClusterTest/java/org/opensearch/analysis/common/QueryStringWithAnalyzersIT.java [hunk 1]
```diff
@@ -49,10 +49,10 @@
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
 
-public class QueryStringWithAnalyzersIT extends ParameterizedOpenSearchIntegTestCase {
+public class QueryStringWithAnalyzersIT extends ParameterizedStaticSettingsOpenSearchIntegTestCase {
```
Comment (reta) [inline modules/analysis-common/src/internalClusterTest/java/org/opensearch/analysis/common/QueryStringWithAnalyzersIT.java:52]: This is a "mechanical" change for ~150 test classes :-\

**Review by peternied — COMMENTED** [review]
Nicely done, looking forward to seeing this PR turn green.

With this change will the `./gradlew ...` output to rerun the test with a seed work?

**Comment by reta** [comment]
> Nicely done, looking forward to seeing this PR turn green.

Thanks @peternied !

> With this change will the ./gradlew ... output to rerun the test with a seed work?

It should, the seed uses whatever `RandomizedContext` is configured with

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java [hunk 2]
```diff
@@ -2270,8 +2122,8 @@ public <T extends TransportRequest> TransportRequestHandler<T> interceptHandler(
      * Returns path to a random directory that can be used to create a temporary file system repo
      */
     public Path randomRepoPath() {
-        if (currentCluster instanceof InternalTestCluster) {
-            return randomRepoPath(((InternalTestCluster) currentCluster).getDefaultSettings());
+        if (cluster() instanceof InternalTestCluster) {
+            return randomRepoPath(((InternalTestCluster) cluster()).getDefaultSettings());
```
[the remark this answers is not in the document]
Comment (reta) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:2126]: This one is a bit tricky - we would be doing `instanceof` several times, here and again in `internalCluster()`

**Comment by neetikasinghal** [comment]
@reta thanks for implementing this.

For classes parameterized with concurrent search setting(which is a dynamic setting), why are we sub-classing them with `ParameterizedStaticSettingsOpenSearchIntegTestCase` as opposed to `ParameterizedDynamicSettingsOpenSearchIntegTestCase`?

**Comment by reta** [comment]
> For classes parameterized with concurrent search setting(which is a dynamic setting), why are we sub-classing them with `ParameterizedStaticSettingsOpenSearchIntegTestCase` as opposed to `ParameterizedDynamicSettingsOpenSearchIntegTestCase`?

@neetikasinghal for two reasons:
 1. The way test suites are run is per parameters combination: all tests with setting = `true`, after all tests with setting = `false` (simple example)
 2. There is no reason to change the setting every single time the tests are run: the setting does not change for whole suite

**Comment by reta** [comment]
> ❌ Gradle check result for [5b6306e](https://github.com/opensearch-project/OpenSearch/commit/5b6306e7d1c660fc6227fbdcf3a501afe66c0f47): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32241/)

https://github.com/opensearch-project/OpenSearch/issues/6090
https://github.com/opensearch-project/OpenSearch/issues/9191
https://github.com/opensearch-project/OpenSearch/issues/10902

**Comment by neetikasinghal** [comment]
> > For classes parameterized with concurrent search setting(which is a dynamic setting), why are we sub-classing them with `ParameterizedStaticSettingsOpenSearchIntegTestCase` as opposed to `ParameterizedDynamicSettingsOpenSearchIntegTestCase`?
> 
> @neetikasinghal for two reasons (and I would argue, it has to be done this way from the beginning):
> 
> 1. The way test suites are run is per parameters combination: all tests with setting = `true`, after all tests with setting = `false` (simple example)
> 2. There is no reason to change the setting every single time the tests are run: the setting does not change for whole suite until parameters are switched over

thanks for explaining, then for what use-cases should `ParameterizedDynamicSettingsOpenSearchIntegTestCase` be used as it can be confusing from its name.

**Comment by reta** [comment]
> thanks for explaining, then for what use-cases should `ParameterizedDynamicSettingsOpenSearchIntegTestCase` be used as it can be confusing from its name.

There are javadocs for each class (and `ParameterizedOpenSearchIntegTestCase`), is it helpful or there are gaps that are unclear? Would appreciate your feedback, thank you.

**Comment by reta** [comment]
> > thanks for explaining, then for what use-cases should `ParameterizedDynamicSettingsOpenSearchIntegTestCase` be used as it can be confusing from its name.
> 
> There are javadocs for each class (and `ParameterizedOpenSearchIntegTestCase`), is it helpful or there are gaps that are unclear? Would appreciate your feedback, thank you.

> also, had another question about what difference would this make when test classes are initialized with @OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST)

Same as on suite level but we will do this for each test:
 - static settings: create cluster -> run test method -> shutdown cluster
 - dynamic settings: create cluster -> apply settings -> run test method -> unapply settings -> shutdown cluster

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (reta) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java]: So for this one, I went ahead and changed `internalCluster()` to return `Optional` instead

File: test/framework/src/main/java/org/opensearch/test/ParameterizedDynamicSettingsOpenSearchIntegTestCase.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.settings.SettingsModule;
+import org.junit.After;
+import org.junit.Before;
+
+/**
+ * Base class for running the tests with parameterization using dynamic settings: the cluster will be created once before the test suite and the
+ * settings will be applied dynamically, please notice that not all settings could be changed dynamically (consider using {@link ParameterizedStaticSettingsOpenSearchIntegTestCase}
+ * instead).
```
Comment (neetikasinghal) [inline test/framework/src/main/java/org/opensearch/test/ParameterizedDynamicSettingsOpenSearchIntegTestCase.java:19]: should we also add a clarifying java doc here stating that - there will be an update to the settings before and after each parameterized run of the test method. So, every test method inheriting ParameterizedDynamicSettingsOpenSearchIntegTestCase  is run number of times equal to the parameters in one go as opposed to `ParameterizedStaticSettingsOpenSearchIntegTestCase` where all the tests with a single parameter are run at a time and then switched over to the other and so one..

**Comment by neetikasinghal** [comment]
> > thanks for explaining, then for what use-cases should `ParameterizedDynamicSettingsOpenSearchIntegTestCase` be used as it can be confusing from its name.
> 
> There are javadocs for each class (and `ParameterizedOpenSearchIntegTestCase`), is it helpful or there are gaps that are unclear? Would appreciate your feedback, thank you.

done

File: modules/analysis-common/src/internalClusterTest/java/org/opensearch/analysis/common/QueryStringWithAnalyzersIT.java (hunk 1 above)
Comment (neetikasinghal) [reply modules/analysis-common/src/internalClusterTest/java/org/opensearch/analysis/common/QueryStringWithAnalyzersIT.java:52]: thanks for taking this up! :)

File: test/framework/src/main/java/org/opensearch/test/ParameterizedDynamicSettingsOpenSearchIntegTestCase.java (hunk 3 above)
Comment (reta) [reply test/framework/src/main/java/org/opensearch/test/ParameterizedDynamicSettingsOpenSearchIntegTestCase.java:19]: Done, thanks @neetikasinghal !

File: test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+          since JUnit creates a new instance per test.
+         */
+        if (suiteInstance != null) {
+            // Catching the case when parameterized test cases are run: the test class stays the same but the test instances changes.
+            if (target instanceof ParameterizedOpenSearchIntegTestCase) {
+                assert suiteInstance instanceof ParameterizedOpenSearchIntegTestCase;
+                if (hasParametersChanged((ParameterizedOpenSearchIntegTestCase) target)) {
+                    printTestMessage("new instance of parameterized test class, recreating cluster scope", method);
+                    afterClass();
+                    beforeClass();
```
Comment (andrross) [inline test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java:369]: Does this logic in effect change a parameterized SUITE scope test into to a TEST scope test? I believe parameterized tests generally run one test method with all parameterized values, then the next test method with all parameterized values, which means every test invocation is going to see a change in parameters here.

**Review by andrross — COMMENTED** [review]
Thanks @reta, in general I think this is great and a better way to structure the test logic within the JUnit framework.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java (hunk 4 above)
Comment (reta) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java:369]: > Does this logic in effect change a parameterized SUITE scope test into to a TEST scope test? 

No, this logic kicks off only when all tests from the sites are executed and new parameter(s) combination is injected so the suite needs to be executed again (a bit more elaborate example of how it works https://github.com/opensearch-project/OpenSearch/pull/11877#issuecomment-1896570247

Comment (andrross) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java:369]: Ah, I was wrong about how the parameterized testing framework orchestrated the tests. Thanks!

**Review by andrross — APPROVED** [review]

File: test/framework/src/main/java/org/opensearch/test/ParameterizedDynamicSettingsOpenSearchIntegTestCase.java (hunk 3 above)
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/ParameterizedDynamicSettingsOpenSearchIntegTestCase.java:19]: Thanks @reta, its much more clearer now

**Review by neetikasinghal — APPROVED** [review]

**Comment by peternied** [comment]
> @peternied how does it look to you?

Thanks for checking in @reta I don't have time for a second pass today, but I'm not concerned about merging with @andrross 's approval.  I'll create issues or a follow up PR if there is anything I'd like to see addressed.

**Comment by Rishikesh1159** [comment]
@reta seems like we are not running all parameter cases of a parameter factory when cluster scope is default or SCOPE.SUITE. Instead we are only running the first case of parameter factory multiple times.

Example [Use Case](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java#L78)

In above example, when we run test class, we are only running the first case of `CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(), false)` twice. This is because we are setting the parameter values in [@override nodeSettings()](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/ParameterizedStaticSettingsOpenSearchIntegTestCase.java#L32). But this method is called only once when the cluster is being [build here](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java#L1949). And in case of `SCOPE.SUITE` we only build cluster once [here](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java#L244). 

So when same test method tries to run with second/different parameters, the node settings are never overridden again with new parameter values because the cluster is already present. Node settings are overridden only when new cluster is built. So this way test is running again with same settings passed in first parameter values.


**Current flow of test runs with multiple parameters and scope set to SCOPE.SUITE or default:**

Build Cluster -> run testMethod1() with param1 -> run testMethod2() with param1 -> run testMethod1() with param1 -> run testMethod2() with param1 -> Remove cluster 

Instead correct flow for using static settings should be : 

Build Cluster -> run testMethod1() with param1 -> run testMethod2() with param1 -> Remove Cluster -> Build cluster again -> run testMethod1() with param2 -> run testMethod2() with param2-> Remove cluster

**Comment by reta** [comment]
@Rishikesh1159 thanks for bringing it up

> So when same test method tries to run with second/different parameters, the node settings are never overridden again with new parameter values because the cluster is already present. Node settings are overridden only when new cluster is built. So this way test is running again with same settings passed in first parameter values.

The cluster should be shut down once the change of parameters is detected, so the new node settings should be picked up, I will double check shortly

**Comment by andrross** [comment]
@Rishikesh1159 `initializeSuiteScope()` should be invoked before each method invocation, and when it detects that parameters have changed it should [rebuild the cluster](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java#L366-L369). Can you look into whether that is happening in the case you're testing?

**Comment by Rishikesh1159** [comment]
@andrross  Yes `initializeSuiteScope()` is being invoked on every test method invocation. But the problem is the conditional check [here](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java#L379C13-L379C30) `if (isSuiteScopedTest(targetClass))`.

We have three type of scopes:
1. @ClusterScope(scope = Scope.SUITE) -  [example](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java#L123)
2. @ClusterScope(scope = Scope.TEST) - [example](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java#L121)
3. @OpenSearchIntegTestCase.SuiteScopeTestCase - [example](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/server/src/internalClusterTest/java/org/opensearch/action/admin/indices/get/GetIndexIT.java#L61)


The [conditional check](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java#L379C13-L379C30) was only detecting 3rd type of scope. But majority of tests are 1st type including this example [use case]([Use Case](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java#L78))

**Comment by reta** [comment]
> @Rishikesh1159 `initializeSuiteScope()` should be invoked before each method invocation, and when it detects that parameters have changed it should [rebuild the cluster](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java#L366-L369). Can you look into whether that is happening in the case you're testing?

@andrross the @Rishikesh1159's explanation (https://github.com/opensearch-project/OpenSearch/pull/11877#issuecomment-1906914937) is correct, I was thinking  exactly the way you described by `initializeSuiteScope()` is not the same as `@ClusterScope(scope = Scope.SUITE)`, it should be fixed by https://github.com/opensearch-project/OpenSearch/pull/12000, thanks folks

**Comment by neetikasinghal** [comment]
> @reta seems like we are not running all parameter cases of a parameter factory when cluster scope is default or SCOPE.SUITE. Instead we are only running the first case of parameter factory multiple times (second case is missing).
> 
> ### **Example [Use Case](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java#L78)**
> In above example, when we run test class, we are only running the first case of `CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(), false)` twice. This is because we are setting the parameter values in [@override nodeSettings()](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/ParameterizedStaticSettingsOpenSearchIntegTestCase.java#L32). But this method is called only once when the cluster is being [build here](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java#L1949). And in case of `SCOPE.SUITE` we only build cluster once [here](https://github.com/opensearch-project/OpenSearch/blob/4167ac5c94c57457bd225fe414f51d1d8fe66ed9/test/framework/src/main/java/org/opensearch/test/OpenSearchTestClusterRule.java#L244).
> 
> So when same test method tries to run with second/different parameters, the node settings are never overridden again with new parameter values because the cluster is already present. Node settings are overridden only when new cluster is built. So this way test is running again with same settings passed in first parameter values.
> 
> ### **Current flow of test runs with multiple parameters and scope set to SCOPE.SUITE or default:**
> Build Cluster -> run testMethod1() with param1 -> run testMethod2() with param1 -> run testMethod1() with param1 -> run testMethod2() with param1 -> Remove cluster
> 
> ### **Instead correct flow for using static settings should be :**
> Build Cluster -> run testMethod1() with param1 -> run testMethod2() with param1 -> Remove Cluster -> Build cluster again -> run testMethod1() with param2 -> run testMethod2() with param2-> Remove cluster

thanks @Rishikesh1159, when I was debugging one of the tests today I realized that even with concurrent search parameter set to true, it wasnt taking the concurrent search path.

**Comment by reta** [comment]
> thanks @Rishikesh1159, when I was debugging one of the tests today I realized that even with concurrent search parameter set to true, it wasnt taking the concurrent search path. This explains on why it was the case.

Sorry about that @neetikasinghal , the issue slipped in

**Pull request merged**